add-go-fuse-to-inject-filesystem-error

qiffang opened the pull request:
**BackGround**
Base the issue - Tests with injected file system errors. #579
We can use go-fuse to inject file system, but it is too complex for the user.

In this pull request, I provide a HOOKFS way.
**Function**
You can implement Hookxxx interface to implement your purposes(For now, only provide HookOnRename interface, if you think it is ok, i will implement other hooks, for example HookOnOpenDir and so on)
```
type HookOnRename interface {
	PreRename(oldPatgh string, newPath string) (hooked bool, err error)
	PostRename(oldPatgh string, newPath string) (hooked bool, err error)
}
```
Then Inject hook implement to HookServer
```
NewHookFs("$orignal", "$mountpoint", &TestRenameHook{})
```

**Example**
```
New HookServer
....
//normal logic
os.Create("$mountpoint/tsdb.txt")
//The rename action in mountpoint path will call Your hook to inject file system errors
err := os.Rename("$mountpoint/tsdb.txt", "$mountpoint/tsdbNew.txt")
```
**QA**
Add test **CreateBlock** logic and **OpenBlock** logic with new way
```
case1:
//1.init hook server, rename files throw IO exception

//2.normal logic
createBlock(t, mountpoint, genSeries(1, 1, 200, 300))
dir, _ := ioutil.ReadDir(mountpoint)
testutil.Equals(t, 0, len(dir))
```

```
case 2:
 //1.create block, this action will be successful because hook server does not start
path := createBlock(t, original, genSeries(1, 1, 200, 300))

//2.init hook server, rename files throw io exception

//3.call OpenBlock method
OpenBlock(nil, filepath.Join(mountpoint, file), nil)
dir, _  := ioutil.ReadDir(filepath.Join(mountpoint, file))
testutil.Equals(t, true, len(dir) > 0)
hasTempFile := false
for _, info := range dir {
	if strings.HasSuffix(info.Name(), "tmp") {
		hasTempFile = true
		break
	}
}

testutil.Equals(t, true, hasTempFile)
```
After all test cases gone, please call "fusermount -u $mountpoint"
There is a complete example in hook_test.go file.
If you think this way is ok, I will implement more hooks.
Thanks

krasi-georgiev commented:
that looks really cool, can you try to implement one full test.

qiffang commented:
> that looks really cool, can you try to implement one full test.

Thanks for your accept, is there a rename relation test case?

krasi-georgiev commented:
the most simple one I can think of is something like:
```
blockDir := createBlock(t, tmpdir, genSeries(1, 1, 0, 0))
b, err := OpenBlock(nil, blockDir, nil)

// check that blockDir doesn't have any tmp files.
```

`OpenBlock` uses `writeMetaFile` which creates a tmp file  and than renames it with the new file so if the rename fails it should leave some tmp files behind.

krasi-georgiev commented:
do you think this will also allow injecting an error on Write? The other bug in the issue I opened is when Write to a file fails due to no disk space #582

qiffang commented:
> the most simple one I can think of is something like:
> 
> ```
> blockDir := createBlock(t, tmpdir, genSeries(1, 1, 0, 0))
> b, err := OpenBlock(nil, blockDir, nil)
> 
> // check that blockDir doesn't have any tmp files.
> ```
> `OpenBlock` uses `writeMetaFile` which creates a tmp file and than renames it with the new file so if the rename fails it should leave some tmp files behind.

I have added two test cases in compact_test.go files, please help to check it.
Thanks

qiffang commented:
> do you think this will also allow injecting an error on `Write`? The other bug in the issue I opened is when `Write` to a file fails due to no disk space #582

Of course, but for now i just implement rename, almost filesystem's action can be injected by this way.
Do not change any real code.

krasi-georgiev commented:
would you mind to fixing travis. Looks like some dependency issues so should be an easy fix.

qiffang commented:
> would you mind fixing the travis tests. Looks like some dependency issues so should be an easy fix.

Currently, the problem is that there is a permission for using fuse

krasi-georgiev commented:
I guess you need to check with the travis team for the proper way to use fuse with travis.

brian-brazil commented:
Fuse is a pretty big requirement for a unittest, that's not going to be runnable in many places.

qiffang commented:
> Fuse is a pretty big requirement for a unittest, that's not going to be runnable in many places.

Thanks for your response, I plan to use it for injecting various file system error in testcases.
Do you think it is ok?
Thanks

krasi-georgiev commented:
before suggesting it as a solution I did check and some travis posts suggest that it should be supported in travis, but of course if there is any other alternative happy to use something easyer.

@simonpasquier @SuperQ maybe you can give some ideas?

qiffang commented:
@krasi-georgiev 
I think we can put the framework to TIDB, but I can remove the test cases util other people fix the permission issue.
Do you think it is ok?

krasi-georgiev commented:
sorry I didn't understand your comment, can you please a bit more details.

qiffang commented:
> sorry I didn't understand your comment, can you please add a bit more details.

@krasi-georgiev 
I mean that can i merge this fuse code without test case?
I remove these test cases for now, do you think it is ok?

krasi-georgiev commented:
we need to be able to run these tests on Travis otherwise why would we need it?

Did you check with the Travis team how to make fuse work there?

krasi-georgiev commented:
ping @qiffang

qiffang commented:
Sorry to reply you so late, but I don't know how to contact Travis team.

krasi-georgiev commented:
https://travis-ci.community/

I would suggest searching within some of existing threads and if can't find anything useful than open a new one.

simonpasquier commented:
Looking at Travis failures, it seems that it works now for Linux and MacOS?
go-fuse doesn't seem to support Windows so you need to add build tags.

krasi-georgiev commented:
That is strange. I think it was failing for all of them before. 
But yeah we don't need it for windows so can exclude it.

qiffang commented:
> That is strange. I think it was failing for all of them before.
> But yeah we don't need it for windows so can exclude it.

I try to find some of existing threads with fuse keyword, but it is empty.
So I have to open a new one, but there is no response for now.

krasi-georgiev commented:
but linux and osx seems to pass now, right? As I mentioned we don't need windows.

krasi-georgiev commented:
aah I see now you have removed the test, that is why it is passing.

can you please enable the test again so that the Travis support can see it failing.

here is the issue you have opened.
https://travis-ci.community/t/fuse-support-in-test-case/3309

qiffang commented:
> but linux and osx seems to pass now, right? As I mentioned we don't need windows.

No, it is also not supported linux environment. And I will enable test cases if Travis response me.

krasi-georgiev commented:
enable it now so they see the failure logs.

krasi-georgiev commented:
also try with 
```
sudo: true
```

krasi-georgiev commented:
also this give some other clues
https://git.drwx.org/bsc/geth/commit/11e7a712f469fb24ddb88ecebcefab6ed8880eb8
```
udo chmod 666 /dev/fuse
sudo chown root:$USER /etc/fuse.conf

```

krasi-georgiev commented:
ping @qiffang

qiffang commented:
> ping @qiffang

Thanks krasi-georgiev , I will try the way which you post.
Thanks so much.

krasi-georgiev commented:
I see you added few more commits, any luck with the suggestions?

qiffang commented:
> I see you added few more commits, any luck with the suggestions?

@krasi-georgiev 
Sorry to reply you so late, I was too busy two days ago

The document which you post is the guide, but we also need to add other commands.

For now, it works well in linux environment.
OS and windows also have problems, i will check it later.
Thanks.

krasi-georgiev commented:
I think we can start with just linux and add osx later if needed.

krasi-georgiev commented:
actually osx  probably not even needed as I would think that most production Prometheus installations would be on Linux with a very small number on Windows.

qiffang commented:
> I think we can start with just linux and add osx later if needed.

MACOS's problem was finally solved, thanks very much for your help.

I have a lot of free time today. I test it one mac and then i found that it is caused by osx’s security policy - “A program tried to load new system extension(s) signed by “Benjamin Fleischer”.  If you want to enable these extensions, open Security & Privacy System Preferences.”

Finally, I found that The xcode9.4 image runs on macOS 10.13, while the xcode8.3 image runs on macOS 10.12. In macOS 10.13, Apple started requiring users to explicitly allow third-party kernel extensions before it will load them.
So Only update image of macOS, it is fixed.

krasi-georgiev commented on .travis.yml:
```diff
@@ -1,20 +1,33 @@
 dist: trusty
 language: go
-os:
-  - windows
```
Sorry maybe I wasn't clear. We still need to run all but the file system  tests for Windows.

For example maybe put all unix OS specific tests in  `block_unix_test.go`

qiffang replied:
> Sorry maybe I wasn't clear. We still need to run all but the file system tests for Windows.
> 
> For example maybe put all unix OS specific tests in `block_unix_test.go`

Currently, go fuse library do not support windows platform, so there is a compile error in windows platform
```
..\..\..\..\pkg\mod\github.com\hanwen\go-fuse@v1.0.0\fuse\types.go:511:2: undefined: Attr
```

I am trying to solve it and do you have any suggestions?
Thanks

krasi-georgiev replied:
why do you think the library  is loaded when running the tests under windows?
When you have something like `block_unix_test.go` the file should be loaded only on unix OS-es

krasi-georgiev replied:
I think you should rename all fuse utils files to fuse_..._unix.go

qiffang replied:
> fuse_..._unix.go

For linux and macos, i add the build flag in golang files
```
// +build linux darwin
```

But for compiling it in windows platform, so there is fuse_xxx_windows.go.
The advantage  is that other developers can write fuse cases in everywhere and do not need to put fuse cases in a special directory.

From the build flag, there is only linux、darwin、windows、freebsd suffix to support different platform. I am not sure the files which suffix is unix used in linux and macos.

krasi-georgiev replied:
Good point, but looking at the code seems that these tests will fail under windows anyway as they will not use the fuse server mounts so  can't inject errors and when the tests expects an error, no error will be returned.

qiffang replied:
@krasi-georgiev 
Currently, for windows platform, it can all cases exclude fuse related cases and for linux and macosx plateform, it can rull all cases. They works well.

I think the tests which need to inject filesystem errors only run in linux and macosx platform.
And in windows platform, it only run other cases.

Do you think it is ok?

krasi-georgiev commented on .travis.yml:
```diff
@@ -1,20 +1,33 @@
+      - sudo addgroup $USER fuse
+      - sudo chmod a+rw /etc/fuse.conf
+      - sudo echo user_allow_other >> /etc/fuse.conf
+      - sudo chmod +x /bin/fusermount
```
Seems strange that this chmod really necessary

qiffang replied:
Yeah, you are right, it is not necessary.
Fixed

qiffang commented:
> #579



> Nice , thanks for the work.
> 
> Can you maybe tidy up the code and make one actual test to cover one of the bugs I mentioned in #579 and I will do an initial review.

Actually, there are 2 test cases in the compact_test.go
1 - TestCreateBlockWithHook
As you know, if we call **createBlock** method, it will call rename method of filesystem.
So before we call this method, we inject a renamed error of filesystem, so there is not files in the dir when we call createBlock method
```
testutil.Equals(t, 0, len(dir))
```
2- TestOpenBlockWithHook
Firstly, it will call **createBlock** method, it is successful.
And before it **OpenBlock** method,  we inject a renamed filesystem error.
So It can not rename meta.json file
```
testutil.Equals(t, false, Exist(filepath.Join(mountpoint, file, "meta.json")))
```

krasi-georgiev commented on .travis.yml:
```diff
@@ -1,20 +1,32 @@
+      - sudo chmod 666 /dev/fuse
+      - sudo chown root:$USER /etc/fuse.conf
+      - sudo addgroup $USER fuse
+      - sudo chmod a+rw /etc/fuse.conf
```
why do we need these?
```
- sudo chmod 666 /dev/fuse
- sudo chown root:$USER /etc/fuse.conf
- sudo chmod a+rw /etc/fuse.conf
```

I think just `- sudo addgroup $USER fuse` should be enough , no?

qiffang replied:
We need to write user_allow_other to fuse.conf to allow other users which is not root to execute fuse

krasi-georgiev replied:
are you sure that `sudo echo user_allow_other >> /etc/fuse.conf` doesn't work without `sudo chmod a+rw /etc/fuse.conf` ?

krasi-georgiev commented on block_test.go:
```diff
@@ -179,8 +179,9 @@ func createBlock(tb testing.TB, dir string, series []Series) string {
-	testutil.Ok(tb, err)
+	ulid, _ := compactor.Write(dir, head, head.MinTime(), head.MaxTime(), nil)
+	//if inject failed operation by fuse, the error is expected
+	//testutil.Ok(tb, err)
```
debug code?

krasi-georgiev commented:
@qiffang did you see my comment at https://github.com/prometheus/tsdb/pull/583#discussion_r289717277

qiffang commented:
> @qiffang did you see my comment at [#583 (comment)](https://github.com/prometheus/tsdb/pull/583#discussion_r289717277)

@krasi-georgiev 
I have responsed you in   [#583 (comment)], for now, every thing works well.
Do you have any suggestions?

qiffang commented:
> @qiffang did you see my comment at [#583 (comment)](https://github.com/prometheus/tsdb/pull/583#discussion_r289717277)

qiffang commented:
> It is almost ready, but needs a lot of tiding up.
> 
> Don't forget that all golang comments should start with a capital letter and end with a full stop.
> Don't forget to check for all errors and fail or log according to each case.
> 
> Once the fuse helper is ready If you allow me I can help with the tests as these need a bit of work as well.

`Don't forget that all golang comments should start with a capital letter and end with a full stop.`
I will update it

`Once the fuse helper is ready If you allow me I can help with the tests as these need a bit of work as well.`
Yes, of course, before we also write our tsdb system. For now, i am very interesting in tsdb. So i want to take part in it. But can i write other cases in another pull request?

qiffang commented on block_test.go:
```diff
@@ -179,8 +179,9 @@ func createBlock(tb testing.TB, dir string, series []Series) string {
-	testutil.Ok(tb, err)
+	ulid, _ := compactor.Write(dir, head, head.MinTime(), head.MaxTime(), nil)
+	//if inject failed operation by fuse, the error is expected
+	//testutil.Ok(tb, err)
```
Done

krasi-georgiev replied:
actually looking at the remaining test you no longer need to check the error so if that is true it doesn't even need the refactoring.

qiffang commented on fileutil/preallocate_darwin.go:
```diff
@@ -34,7 +34,7 @@ func preallocFixed(f *os.File, sizeInBytes int64) error {
 	p := unsafe.Pointer(fstore)
 	_, _, errno := syscall.Syscall(syscall.SYS_FCNTL, f.Fd(), uintptr(syscall.F_PREALLOCATE), uintptr(p))
-	if errno == 0 || errno == syscall.ENOTSUP {
+	if errno == 0 || errno == syscall.ENOTSUP || errno == syscall.EINVAL {
```
I used an old way so it will has ENOTSUP error, but for now we do not need it.
Deleted it, thanks

qiffang commented on testutil/fuse_utils.go:
```diff
@@ -0,0 +1,61 @@
+	"testing"
+)
+
+func NewFuseServer(t *testing.T, original, mountpoint string, hook Hook) (interface{}, error) {
```
For making the same interface with fuse_utils_windows.go, Windows plateform has no fuse function, there will be compile error if we import fuse package in compact_test.go file.

krasi-georgiev replied:
There should be another way to achieve this. ping if you can't figure it out and I can have a look, but using an empty interface should be used only as a last resort.

qiffang replied:
> There should be another way to achieve this. ping if you can't figure it out and I can have a look, but using an empty interface should be used only as a last resort.

I will try it, thanks

qiffang replied:
> There should be another way to achieve this. ping if you can't figure it out and I can have a look, but using an empty interface should be used only as a last resort.

Currently, I add a new struct named Server, please help to check it.
Thanks

krasi-georgiev replied:
yes I think that should be fine, but will have another look when I do the next review.

qiffang commented on testutil/fuse_utils.go:
```diff
@@ -0,0 +1,61 @@
+	return server, nil
+}
+
+func CleanUp(s interface{}, mountpoint string, original string) {
```
At the beginning, i write as your suggestions. But fuse package has not windows function, so there will be compile error if we import fuse package in compact_test.go file.

krasi-georgiev replied:
in that case try the other suggest to build this only for `linux and darwin`.

qiffang replied:
As you said that we need to run other cases in windows platform by travis-ci

qiffang replied:
Updated

qiffang commented on testutil/fuse_utils.go:
```diff
@@ -0,0 +1,61 @@
+
+	os.RemoveAll(mountpoint)
+	os.RemoveAll(original)
+	log.Info("Done")
```
Removed

qiffang commented on testutil/fuse_utils.go:
```diff
@@ -0,0 +1,61 @@
+func createDirIfAbsent(name string) {
+	_, err := os.Stat(name)
+	if err != nil {
+		os.Mkdir(name, os.ModePerm)
```
If the directory exist, there will be error.
os.Stat will has error if the directory existed.

krasi-georgiev replied:
according to the golang doc it returns nil if the dir already exists.
https://golang.org/pkg/os/#MkdirAll

qiffang replied:
Updated

qiffang commented on compact_test.go:
```diff
@@ -1055,3 +1057,58 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
 	}
 }
+
+func TestCreateBlockWithHook(t *testing.T) {
```
Removed

krasi-georgiev commented:
> Yes, of course, before we also write our TSDB system. I am very interesting in TSDB. So i want to take part in it. But can i write other cases in another pull request, else there are too many code chages because there are some methods to implement, for example delete directory, delete files, list directory, list files and son on?

if you are so interested I will leave the tests to you and will give directions. Finish the proposed changes and ping so I can let you know about the test code change suggestions.

qiffang commented on testutil/fs_hook.go:
```diff
@@ -0,0 +1,210 @@
+// +build linux darwin
```
Done

qiffang commented on testutil/fs_hook.go:
```diff
@@ -0,0 +1,210 @@
+	Mountpoint string
+	FsName     string
+	fs         pathfs.FileSystem
+	hook       Hook
```
Because there will be many Hook type in the future.

For example:
```
HookOnOpen //called on open
HookOnRead // called on read
HookOnWrite //called on write
HookOnMkdir //called on mkdir
HookOnRmdir 
HookOnOpenDir
HookOnFsync
...
```
They are also implements Hook. As i said, in this pull request, I just implement fuse framework and i will add all fuse support in another pull request.

krasi-georgiev replied:
ok, noted, but lets try with an exact type and will think about the other ones with the other PR.

qiffang replied:
It is an obviously interface, do you think it is very necessary to update Hook to HookOnRename?

krasi-georgiev replied:
Yes I think typecasting is error prone and should use it at last resort.

krasi-georgiev replied:
I have this idea using composition:

```

type Hook interface {
	PreRename(oldPatgh string, newPath string) (hooked bool, err error)
	PostRename(oldPatgh string, newPath string) (hooked bool, err error)
	Chmod(name string, mode uint32, context *fuse.Context) fuse.Status
	.
	.
	. // All other funcs from the fuse server.
}

type TestRenameHook struct {
	testutil.EmptyHook // Add the Empty hook so the that this struct implements the hook interface.
}

// These will take precedence over the `testutil.EmptyHook`
func (h TestRenameHook) PreRename(oldPatgh string, newPath string) (hooked bool, err error) {
	fmt.Printf("renamed file from %s to %s \n", oldPatgh, newPath)
	return true, syscall.EIO
}
func (h TestRenameHook) PostRename(oldPatgh string, newPath string) (hooked bool, err error) {
	return false, nil
}


type EmptyHook struct{}

func (h EmptyHook) PreRename(oldPatgh string, newPath string) (hooked bool, err error) {
	return false, nil
}
func (h EmptyHook) PostRename(oldPatgh string, newPath string) (hooked bool, err error) {
	return false, nil
}
func (h EmptyHook) Chmod(name string, mode uint32, context *fuse.Context) fuse.Status { return fuse.Status(0) }
.
.
.
// All other funcs from the Hook interface.

```

qiffang commented on testutil/hook.go:
```diff
@@ -0,0 +1,25 @@
+// +build linux darwin
```
As i described in [above](https://github.com/prometheus/tsdb/pull/583#discussion_r291123325), there will be many hooks

krasi-georgiev replied:
Maybe it will still remain relatively small, but we can split later if needed. For now lets keep the minimum files required.

qiffang replied:
Ok, Removed it to fuse.go file

qiffang commented:
> I will leave the tests to you and will give directions. Finish the proposed changes and ping so I can let you know about the test code change suggestions

No problem. In Logicmonitor, we have own TSDB system because of we provide a SASS platform. The data is very large. After I finish all test cases, I hope I can take part in development coding.

krasi-georgiev commented on testutil/fuse.go:
```diff
@@ -0,0 +1,250 @@
+
+	//async start fuse server
+	go func() {
+		fs.Start(server)
```
where is this stopped ?

qiffang replied:
> where is this stopped ?

If call `syscall.Unmount`, it will stop. And it will be called in CleanUp method

krasi-georgiev replied:
thanks, could you put a comment to clarify this.

qiffang replied:
Done

krasi-georgiev commented on block_test.go:
```diff
@@ -150,16 +157,15 @@ func TestCorruptedChunk(t *testing.T) {
 
 // createBlock creates a block with given set of series and returns its dir.
-func createBlock(tb testing.TB, dir string, series []Series) string {
+func createBlock(tb testing.TB, dir string, series []Series) (string, error) {
```
looking at the changes you don't need to refactor this anymore.

krasi-georgiev replied:
I meant it can stay as before:
`func createBlock(tb testing.TB, dir string, series []Series) string {`

qiffang replied:
> I meant it can stay as before:
> `func createBlock(tb testing.TB, dir string, series []Series) string {`
> 
> if it is not needed in this PR can refactor it when needed.

Restore it as before

krasi-georgiev commented on testutil/fuse.go:
```diff
@@ -0,0 +1,261 @@
+	server.Serve()
+}
+
+func newHookServer(hookfs *HookFs) (*fuse.Server, error) {
```
I don't see why is this a seperate func?

krasi-georgiev commented on compact_test.go:
```diff
@@ -1058,3 +1065,45 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
+	path, err := createBlock(t, original, genSeries(1, 1, 200, 300))
+	testutil.Ok(t, err)
+	_, file := filepath.Split(path)
+	server, err := testutil.NewServer(t, original, mountpoint, testutil.Hook(TestRenameHook{}))
```
`testutil.NewServer` it is a bit unclear what is this new server. Best to have it as before `testutil.NewFuseServer` or maybe put as a child package of `testutil/fuse` so that this becomes `fuse.NewServer`

qiffang replied:
Acceptd and Updated

qiffang commented on testutil/fuse_windows.go:
```diff
@@ -0,0 +1,30 @@
+// +build windows
```
`we don't need this file anymore right?`
Yes, removed

krasi-georgiev replied:
No I meant is there no way to completely remove the file?

qiffang replied:
> No I meant is there no way to completely remove the file?

But for the fuse library, it also use build flag ` +build linux darwin` to distinct different platform.
For windows platform, there is no the function which will trigger compile error.

Can you explain why we do not use this way?
Thanks

qiffang commented:
Hi @krasi-georgiev 
Thanks for your [comment](https://github.com/prometheus/tsdb/pull/583#discussion_r291950128)

But if there are some limitations as your design, as you see in the fuse.go
```
func (h *HookFs) Rename(oldName string, newName string, context *fuse.Context) fuse.Status {
	hook, hookEnabled := h.hook.(HookOnRename)
	if hookEnabled {
		preHooked, err := hook.PreRename(oldName, newName)
		if preHooked {
			if err != nil {

				return fuse.ToStatus(err)
			}
		}
	}

	status := h.fs.Rename(oldName, newName, context)

	if hookEnabled {
		postHooked, err := hook.PostRename(oldName, newName)
		if postHooked {
			if err != nil {
				return fuse.ToStatus(err)
			}
		}
	}

	return status
}
```
In different method, we need to distinct the hook type.
But as my design, it allows many TestRenameHook struct because different cases need to do the different things in `PreRename` and `PostRename` method.

qiffang commented:
It is a good [design](https://github.com/prometheus/tsdb/pull/583#discussion_r291950128), and thanks very much.
Updated

krasi-georgiev commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,46 @@
+
+// These will take precedence over the `testutil.EmptyHook`
+func (h TestRenameHook) PreRename(oldPatgh string, newPath string) (hooked bool, err error) {
+	fmt.Printf("renamed file from %s to %s \n", oldPatgh, newPath)
```
debug code?

qiffang replied:
Removed

krasi-georgiev commented on compact_test.go:
```diff
@@ -1058,3 +1061,33 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
+func TestOpenBlockWithHook(t *testing.T) {
+	//init action
+	original := filepath.Join(string(filepath.Separator), "tmp", fmt.Sprintf("dev-%d", time.Now().Unix()))
+	mountpoint := filepath.Join(string(filepath.Separator), "tmp", fmt.Sprintf("mountpoint-%d", time.Now().Unix()))
```
something like this would be a bit better
```
original, err := ioutil.TempDir("", "test")
testutil.Ok(t, err)
mountpoint, err := ioutil.TempDir("", "test")
testutil.Ok(t, err)
```

than can remove the creation logix in the NewServer.

qiffang replied:
Accept and updated

qiffang commented:
```
are you sure that sudo echo user_allow_other >> /etc/fuse.conf doesn't work without sudo chmod a+rw /etc/fuse.conf ?
```
Yes, perviously, there is only `sudo echo user_allow_other >> /etc/fuse.conf`, there is permission error

qiffang commented:
> I just tried this locally on Ubuntu 18.10 and I gets stuck with
> 
> ```
> go test -timeout 60s github.com/prometheus/tsdb -run ^TestOpenBlockWithHook$ -v
> === RUN   TestOpenBlockWithHook
> 2019/06/11 17:16:13 Unimplemented opcode POLL
> --- PASS: TestOpenBlockWithHook (0.38s)
> PASS
> ```

Which go version you are using? can you check the configuration with travis.yaml?

qiffang commented on querier_test.go:
```diff
@@ -1523,6 +1523,7 @@ func BenchmarkQueryIterator(b *testing.B) {
 					} else {
 						generatedSeries = populateSeries(prefilledLabels, mint, maxt)
 					}
+
```
Done

qiffang commented:
```
are you sure that sudo echo user_allow_other >> /etc/fuse.conf doesn't work without sudo chmod a+rw /etc/fuse.conf ?
```
Hi @krasi-georgiev , If I removed `sudo chmod a+rw /etc/fuse.conf`
![image](https://user-images.githubusercontent.com/7285248/59331252-101b2980-8d26-11e9-98a8-4508ca221ef0.png)

There is a permission issue as i said before.

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -236,9 +236,31 @@ func NewServer(t *testing.T, original, mountpoint string, hook Hook) (*Server, e
-	syscall.Unmount(s.mountpoint, -1)
+	if err := s.server.Unmount(); err != nil {
+		if err = s.forceMount(); err != nil {
+			fmt.Println("Umount failed", fmt.Sprintf("mountpoint=%s, err=%v", s.mountpoint, err))
```
1. Why the program get stuck
Actually, the tests is finished.
And then we will umount the mountpoint, but in my platform(osx), i call `syscall.Unmount(s.mountpoint,  -1)`, but -1 is invalid argument in linux platform.
The flag should be `0x1` in linux platform for the `Unmount` method

2. Why I just print the log if the unmount failed
The unmount failed is not test failed, and we can assure that the unmount must be successful.

krasi-georgiev replied:
hm we need to investigate this as even though the test is not failed I think it makes sense to fail even in cases where the cleanup fails.

candlerb commented:
@qiffang:

> Yes, perviously, there is only `sudo echo user_allow_other >> /etc/fuse.conf`, there is permission error

That's because the redirection of the file (`>> /etc/fuse.conf`) happens in the shell before sudo is run, so is done as a regular user.

The canonical way to fix this is with `tee`:

```
echo user_allow_other | sudo tee -a /etc/fuse.conf
```

Or to run a subshell in sudo:

```
sudo sh -c 'echo user_allow_other >>/etc/fuse.conf'
```

krasi-georgiev commented:
> sudo sh -c 'echo user_allow_other >>/etc/fuse.conf'

this one looks ok.

qiffang commented:
> > sudo sh -c 'echo user_allow_other >>/etc/fuse.conf'
> 
> this one looks ok. and maybe add a comment why it is like this.

Ok,  but for now, i am investigating why the program maybe failed when unmount  the directory when we call `CleanUp` method even though the test is not failed.

qiffang commented:
> @qiffang:
> 
> > Yes, perviously, there is only `sudo echo user_allow_other >> /etc/fuse.conf`, there is permission error
> 
> That's because the redirection of the file (`>> /etc/fuse.conf`) happens in the shell before sudo is run, so is done as a regular user.
> 
> The canonical way to fix this is with `tee`:
> 
> ```
> echo user_allow_other | sudo tee -a /etc/fuse.conf
> ```
> 
> Or to run a subshell in sudo:
> 
> ```
> sudo sh -c 'echo user_allow_other >>/etc/fuse.conf'
> ```

Thanks @candlerb . Yes, you are right, it can write `user_allow_other` to fuse.conf file.
But i forgot to say that `fusermount` process which we will use it in the program also use `fuse.conf` file. If I do not give the permission to `fuse.conf`, there also be permission issue.
```
>> running all tests
GO111MODULE=on go test -race  ./...
/bin/fusermount: failed to open /etc/fuse.conf: Permission denied
```

qiffang commented:
> > sudo sh -c 'echo user_allow_other >>/etc/fuse.conf'
> 
> this one looks ok. and maybe add a comment why it is like this.

Hi @krasi-georgiev 
This way can write `user_allow_other ` to fuse.conf file. But `fusermount` process which we will use it in the program also use `fuse.conf` file. If I do not give the permission to fuse.conf, there also be permission issue.
```
>> running all tests
GO111MODULE=on go test -race  ./...
/bin/fusermount: failed to open /etc/fuse.conf: Permission denied
```

For now, all issues have been fixed, please try it again in your environment with the latest code  @krasi-georgiev .

Thanks

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,238 @@
+		h.Original, h.Mountpoint, h.FsName, h.fs.String(), h.hook)
+}
+
+func (h *HookFs) SetDebug(debug bool) {
```
I just tried it and you can avoid all this extra boilerplate by using struct embedding like:

```
type LoopBackFs struct {
	pathfs.FileSystem
}

type HookFs struct {
	Original   string
	Mountpoint string
	FsName     string
	LoopBackFs
	hook Hook
}

func NewHookFs(original string, mountpoint string, hook Hook) (*HookFs, error) {
	hookfs := &HookFs{
		Original:   original,
		Mountpoint: mountpoint,
		FsName:     "hookfs",
		LoopBackFs: LoopBackFs{pathfs.NewLoopbackFileSystem(original)},
		hook:       hook,
	}
	return hookfs, nil
}
```

qiffang replied:
Thanks for your comments, and updated it.
But for now, there is only rename which has hook, actually almost all methods should have hook to inject all filesystem error.
So if almost methods have the hook like rename method, there will be also `LoopBackFs` boilerplate.

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,238 @@
+		}
+	}
+
+	status := h.fs.Rename(oldName, newName, context)
```
after the struct embedding this will become:
```suggestion
	status := h.LoopBackFs.Rename(oldName, newName, context)
```

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,238 @@
+	return server, nil
+}
+
+//tests will want to run this in a  goroutine.
```
I think I mentioned it few times, but all comments should start with a capital letter and end with a full stop.

Also all exported objects should have a comment following the golang rules.

qiffang replied:
Thanks for your patience,  I will bear it in mind. I used golang language for 3 months and I learned a lot from you. Thanks very much.

qiffang replied:
Done, please review it again.
Thanks

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,130 @@
+
+// A FUSE filesystem that shunts all request to an underlying file system.
+// Its main purpose is to provide test coverage.
+type hookFs struct {
```
look here for documenting exported and unexperted objects
blog.golang.org/godoc-documenting-go-code

the comment should always start with the name of the object
`// hookFs .....`

this is because of the auto generate docs.

qiffang replied:
Accept and I will updated it.
Thanks

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,130 @@
+	hook Hook
+}
+
+func newHookFs(original string, mountpoint string, hook Hook) (*hookFs, error) {
```
don't think we need this as separate func for now.

qiffang replied:
Accept and I will updated it.

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,130 @@
+		server.Serve()
+	}()
+
+	testutil.Ok(t, server.WaitMount())
```
if this errors - `Unmount()` will never be called and the test will probably hang.

qiffang replied:
As your suggestion, there will not be  `testing` in `NewServer` function.
I will return error for `NewServer` function, and test case should call `defer close()`

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,130 @@
+		}
+
+		testutil.Ok(t, os.RemoveAll(mountpoint))
+		testutil.Ok(t, os.RemoveAll(original))
```
These are created in the test so best to do the removal in the test as well.

qiffang replied:
Accept and I will updated it.

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,130 @@
+	testutil.Ok(t, err)
+
+	server, err := fs.newServer()
+	testutil.Ok(t, err)
```
I think it is best that when there is an error to just skip the test.
This will not enforce users to have a fuse server installed locally to run tests and we will make sure the these are not skipped during CI.

Or if there is any other way to detect when fuse server is not installed, show a warning and continue with the other tests.

qiffang replied:
There is a fast fail if fuse  is not installed.
I think whether skip the error should be controlled by test case.
So if test case call `fuse.NewServer` and this function return error,  the test case decide to skip it.
Do you think it is ok?

krasi-georgiev replied:
I didn't quite understand the, idea, can you maybe show it with some code.

qiffang replied:
With the new code, if the custom do not install fuse locally, the test case will skip tests.
I run it in a machine which do not install fuse locally, the result is here:
```
[root@172.16.4.65 tsdb]# go test -timeout 60s github.com/prometheus/tsdb -run ^TestFailedDelete$ -v

=== RUN   TestFailedDelete
--- SKIP: TestFailedDelete (0.36s)
    compact_test.go:1078: use server couldn't be started
PASS
ok  	github.com/prometheus/tsdb	0.383s
```

krasi-georgiev commented on compact_test.go:
```diff
@@ -1058,3 +1060,38 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
+
+	// Create block will be successful because hook server does not start.
+	_, file := filepath.Split(createBlock(t, original, genSeries(1, 1, 200, 300)))
+	clean := fuse.NewServer(t, original, mountpoint, fuse.Hook(fuse.TestRenameHook{}))
```
instead of `clean` you can have `func (h *hookFs) Close() error { return server.Unmount() }` so it becomes like 

```
server,error := fuse.NewServer(original, mountpoint, fuse.FailingRenameHook{})
testutil.Ok(t, err)
defer func(){
	testutil.Ok(t, server.Close())
}()
```

Note that I have taken the t testing.T out of the NewServer arguments. best to keep this in the tests only.

Note that you don't need casting for `fuse.Hook(fuse.FailingRenameHook{}) -> fuse.FailingRenameHook{}`

qiffang replied:
Accept and I will updated it.

@krasi-georgiev , is it convenient to add a contact information?

qiffang replied:
Hi @krasi-georgiev 
If we take the t testing.T out of the NewServer arguments, the test cases will like this:
```
	server, err := fuse.NewServer(original, mountpoint, fuse.FailingRenameHook{})
	// Remember to call unmount after you do not use it.
	defer func() {
		if server != nil {
			testutil.Ok(t, server.Close())
		}
	}()

	if err != nil {
		t.Skip()
	}
```
These codes will be redundant, i think it is better to do this check only in NewServer method, not in every test cases.

krasi-georgiev replied:
> is it ok to add a contact information?

not sure what you mean by this.


> These codes will be redundant, i think it is better to do this check only in NewServer method, not in every test cases.

It is fine to do this check with every test. Even with a bit of repetitive code is better for being more explicit and show why the test is skipped instead if hiding it in `fuse.NewServer`



this is how I see the final result.

```
server, err := fuse.NewServer(original, mountpoint, fuse.FailingRenameHook{})
if err != nil {
	t.Skip("use server couldn't be started")
}	
defer func() {
	testutil.Ok(t, server.Close())
}()

```

qiffang commented:
> Yes it works now.
> 
> I have refactored the test and even found a bug which I will fix in a seprate PR.
> 
> ```
> // TestFailedDelete ensures that the block is in its original state when a delete fails.
> func TestFailedDelete(t *testing.T) {
> 	original, err := ioutil.TempDir("", "original")
> 	testutil.Ok(t, err)
> 	mountpoint, err := ioutil.TempDir("", "mountpoint")
> 	testutil.Ok(t, err)
> 
> 	defer func(){
> 		testutil.Ok(t, os.RemoveAll(mountpoint))
> 		testutil.Ok(t, os.RemoveAll(original))
> 	}()
> 
> 	_, file := filepath.Split(createBlock(t, original, genSeries(1, 1, 1, 100)))
> 	clean := fuse.NewServer(t, original, mountpoint, fuse.FailingRenameHook{})
> 	defer clean()
> 
> 	expHash, err := testutil.DirHash(original)
> 	testutil.Ok(t, err)
> 	fmt.Println("-------")
> 	pb, err := OpenBlock(nil, filepath.Join(mountpoint, file), nil)
> 	testutil.Ok(t, err)
> 	defer func() {
> 		testutil.Ok(t, pb.Close())
> 	}()
> 
> 	testutil.NotOk(t, pb.Delete(1, 10, labels.NewMustRegexpMatcher("", ".*")))
> 
> 	actHash, err := testutil.DirHash(original)
> 	testutil.Ok(t, err)
> 
> 	testutil.Equals(t, expHash, actHash, "the block dir hash has changed after a failed delete")
> 
> }
> ```
> 
> ../testutil/directory.go
> 
> ```
> // DirHash returns a hash of all files attributes and their content within a directory.
> func DirHash(path string) ([]byte, error) {
> 	hash := md5.New()
> 	err := filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
> 		if err != nil {
> 			return err
> 		}
> 		if !info.IsDir() {
> 			f, err := os.Open(path)
> 			if err != nil {
> 				return err
> 			}
> 			defer f.Close()
> 
> 			if _, err := io.Copy(hash, f); err != nil {
> 				return err
> 			}
> 
> 			if _, err := io.WriteString(hash, strconv.Itoa(int(info.Size()))); err != nil {
> 				return err
> 			}
> 			if _, err := io.WriteString(hash, info.Name()); err != nil {
> 				return err
> 			}
> 			modTime, err := info.ModTime().GobEncode()
> 			if err != nil {
> 				return err
> 			}
> 			if _, err := io.WriteString(hash, string(modTime)); err != nil {
> 				return err
> 			}
> 
> 		}
> 		return err
> 	})
> 	return hash.Sum(nil), err
> }
> ```

This case fail for now, so i have to wait [#636](https://github.com/prometheus/tsdb/pull/636)

qiffang commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,44 @@
+	"syscall"
+)
+
+type Hook interface {
```
Add more functions in Hook and EmptyHook, but they are not called in fuse.

qiffang commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,44 @@
+	PostRename(oldPatgh string, newPath string) (hooked bool, err error)
+}
+
+type TestRenameHook struct {
```
Done

krasi-georgiev commented:
> This case fail for now, so i have to wait #636?

yes

qiffang commented:
> > This case fail for now, so i have to wait #636?
> 
> yes

Got it and do you have any suggestions for this pull request?

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+		h.original, h.mountpoint, h.fsName, h.loopBackFs.String(), h.hook)
+}
+
+// Rename implements hanwen/go-fuse/fuse/pathfs.FileSystem. You are not expected to call h manually.
```
It is most common to note that a given struct implements an interface and the describe the actual funcs closer to what they actually do.

```suggestion
// Rename calls the injected rename hooks if those exist or the underlying loopback fs Rename api.
```

qiffang replied:
Got it, Done

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	return fuse.NewServer(conn.RawFS(), h.mountpoint, mOpts)
+}
+
+// Server is proxy of fuse server.
```
```suggestion
// Server is a fuse server proxy.
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// NewServer creates a fuse server and attaches it to the given `mountpoint` directory.
+// It returns a function to `unmount` the given `mountpoint` directory.
```
```suggestion
// The caller should not forget to close the server to release the mountpoints.

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	}, server.WaitMount()
+}
+
+// Close return false if unmount `mountpoint` faild. But if the caller open files but forget to close them,
```
```suggestion
// Close releases the mountpoints and return false if the unmount fails.
// When the mountpoint has open files it tries to force unmount.
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// Close return false if unmount `mountpoint` faild. But if the caller open files but forget to close them,
+// we also force unmout them to avoid test case stuck.
```
```suggestion
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+// Copyright 2019 The qiffang Authors
```
```suggestion
// Copyright 2019 The Prometheus Authors
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
you still need `// +build windows` no?

qiffang replied:
Naming a file fuse_windows.go will cause it to be included only when building it for Windows, it is the same as adding the build flag "// +build !linux,!darwin !cgo" in the file.

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+
+import "errors"
+
+// Server is empty struct because windows platform do not support fuse.
```
```suggestion
// Server implements a mock server to allow skipping test for windows platform as it doesn't support fuse.
```

qiffang replied:
Done

krasi-georgiev replied:
There was a typo and maybe add a new line:

```
// Server implements a mock server to allow skipping tests for windows platform 
// which doesn't support fuse.

```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+type Server struct {
+}
+
+// NewServer is a mock method because fuse does not support Windows platform.
```
```suggestion
// NewServer always returns an error because fuse does not support Windows platform.
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+	return nil, errors.New("Unsupported fuse.")
+}
+
+// Close is a mock method
```
```suggestion
// Close is a noop method
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+// Close is a mock method
+func (s *Server) Close() (err error) {
```
```suggestion
func (s *Server) Close() error {
```

qiffang replied:
Done

krasi-georgiev commented on compact_test.go:
```diff
@@ -1058,3 +1059,40 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
+	_, file := filepath.Split(createBlock(t, original, genSeries(1, 1, 1, 100)))
+	server, err := fuse.NewServer(original, mountpoint, fuse.FailingRenameHook{})
+	if err != nil {
+		t.Skip("use server couldn't be started")
```
```suggestion
		t.Skip(err) // Skip the test for any error. These tests are optional
```

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/fuse_windows.go:
```diff
@@ -0,0 +1,29 @@
+
+// NewServer is a mock method because fuse does not support Windows platform.
+func NewServer(original, mountpoint string, hook Hook) (*Server, error) {
+	return nil, errors.New("Unsupported fuse.")
```
errors strings start with lower case and should NOT end in full stop.
```suggestion
	return nil, errors.New("fuse not supported under Windows")
```

I recommend install vscode as it gives you hints for all these golang idioms.

qiffang replied:
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+func (s *Server) forceMount() (err error) {
+	delay := time.Duration(0)
+	for try := 0; try < 5; try++ {
+		err = syscall.Unmount(s.mountpoint, flag)
```
If we write the code like `syscall.Unmount(s.mountpoint, unmountFlag())`, it have to call `unmountFlag` method every time when the forceMount is called.  Actually, the `unmountFlag` only should be computed once.  So I'm introducing a constant `flag`

krasi-georgiev replied:
There is no problem to call unmountFlag() few times. This is no requirement for high performance here and I think the compiler would be smart enough to translate this into a single call.

qiffang commented on compact_test.go:
```diff
@@ -1058,3 +1059,40 @@ func TestDeleteCompactionBlockAfterFailedReload(t *testing.T) {
 }
+
+// TestFailedDelete ensures that the block is in its original state when a delete fails.
+func TestFailedDelete(t *testing.T) {
```
Done

qiffang commented on testutil/testutil.go:
```diff
@@ -85,3 +89,41 @@ func formatMessage(msgAndArgs []interface{}) string {
 }
+
+// DirHash returns a hash of all files attributes and their content within a directory.
+func DirHash(path string) ([]byte, error) {
```
Done

qiffang commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,167 @@
+	return false, nil
+}
+
+// FailingRenameHook is called on rename directory.
```
Done

qiffang commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,167 @@
+	EmptyHook
+}
+
+// PreRename will take precedence over the `EmptyHook.PreRename()`.
```
Done

qiffang commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,167 @@
+	return true, syscall.EIO
+}
+
+// PostRename will take precedence over the `EmptyHook.PostRename()`.
```
Done

qiffang commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 The qiffang Authors
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+// +build linux darwin
+
+// Copyright 2019 The qiffang Authors
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// hookFs filesystem that shunts all request to an underlying file system.
+// Its main purpose is to provide test coverage.
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	hook Hook
+}
+
+// String implements hanwen/go-fuse/fuse/pathfs.FileSystem. You are not expected to call h manually.
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// forceMount force to unmount `mountpoint` to avoid the case that caller open files but forget to close them.
+func (s *Server) forceMount() (err error) {
```
Thanks, done.

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	return
+}
+
+// forceMount force to unmount `mountpoint` to avoid the case that caller open files but forget to close them.
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+// we also force unmout them to avoid test case stuck.
+func (s *Server) Close() (err error) {
+	if err = s.server.Unmount(); err != nil {
+		s.forceMount()
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	flag = unmountFlag()
+)
+
+// unmountFlag return force unmount flag based different platform.
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	pathfs.FileSystem
+}
+
+// hookFs filesystem that shunts all request to an underlying file system.
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+	preHooked, err := h.hook.PreRename(oldName, newName)
+	if preHooked {
+		if err != nil {
+
```
Done

qiffang commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,164 @@
+
+// Close return false if unmount `mountpoint` faild. But if the caller open files but forget to close them,
+// we also force unmout them to avoid test case stuck.
+func (s *Server) Close() (err error) {
```
Done

krasi-georgiev commented on block_test.go:
```diff
@@ -236,3 +238,40 @@ func populateSeries(lbls []map[string]string, mint, maxt int64) []Series {
+	_, file := filepath.Split(createBlock(t, original, genSeries(1, 1, 1, 100)))
+	server, err := fuse.NewServer(original, mountpoint, fuse.FailingRenameHook{})
+	if err != nil {
+		t.Skip(err) // Skip the test for any error. These tests are optional
```
```suggestion
		t.Skip(err) // Skip the test for any error. These tests are optional.
```

qiffang replied:
Done

krasi-georgiev commented:
#636 is merged so you can merge it here to pass the test.

qiffang commented:
```
There is no problem to call unmountFlag() few times. This is no requirement for high performance here and I think the compiler would be smart enough to translate this into a single call.
```

For golang, it does not have JIT(Java has it, so java complier can translate this into a single call), golang just put it as inline function. I think it should be fast depend on cpu branch prediction.

As you said, it is no requirement for high performance here and i will remove the flag variables.

krasi-georgiev commented on testutil/fuse/fuse.go:
```diff
@@ -0,0 +1,160 @@
+	pathFsOpts := &pathfs.PathNodeFsOptions{ClientInodes: true}
+	pathFs := pathfs.NewPathNodeFs(h, pathFsOpts)
+	conn := nodefs.NewFileSystemConnector(pathFs.Root(), opts)
+	originalAbs, _ := filepath.Abs(h.original)
```
unhandled error.

qiffang replied:
Done

krasi-georgiev commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,167 @@
+
+// FailingRenameHook implements the hook interface and fails on renaming operations.
+type FailingRenameHook struct {
+	// Add the Empty hook so the that this struct implements the hook interface.
```
to follow the golang comment  standards.
```suggestion
	// EmptyHook is embedded so that the FailingRenameHook fully implements the hook interface.
```

qiffang replied:
Oh my god, i just copied your comments here, i am so sorry.
Done

krasi-georgiev commented on testutil/fuse/hook.go:
```diff
@@ -0,0 +1,167 @@
+
+// FailingRenameHook implements the hook interface and fails on renaming operations.
+type FailingRenameHook struct {
+	// EmptyHook is embedded so that the FailingRenameHook fully struct implements the hook interface.
```
```suggestion
	// EmptyHook is embedded so that the FailingRenameHook fully implements the hook interface.
```

qiffang replied:
Done

krasi-georgiev commented:
@codesome 	, @bwplotka this is ready for your review.

long story short is that may bugs are caused by some file system errors and with this Pr we will be able to write tests that inject fs errors and ensure that tsdb database state is left in a recoverable state. Since this uses fuse the tests are optional and meant to run mainly on the CI and targeting linux.

krasi-georgiev commented:
@qiffang while waiting for a review do you want to write a test for #582

Something like - 
1. start creating  WAL records
1. after few records  inject an error the the file system is full.
1. check that the `appender.Add` fails, 
1. Remove the file system is full injection error.
1. check that `appender.Add` succeeds.

qiffang commented:
> @qiffang while waiting for a review do you want to write a test for #582.
> It is fine if you want to make this here or in a separate PR after/if this one is merged.
> 
> Something like -
> 
> 1. start creating  WAL records
> 2. after few records  inject an error the the file system is full.
> 3. check that the `appender.Add` fails,
> 4. Remove the file system is full injection error.
> 5. check that `appender.Add` succeeds.

Hi @krasi-georgiev , I think it is better to do it in a separate PR.
For now, i will write it in our local environment and i will submit it after this PR is merged.
Thanks

krasi-georgiev commented:
ping @bwplotka 	@codesome 	for a review.

krasi-georgiev commented:
@qiffang failing tests.

@bwplotka , @codesome 	, @gouthamve 	would you minf having a look at this PR.

krasi-georgiev commented on block_test.go:
```diff
@@ -316,7 +318,7 @@ func TestFailedDelete(t *testing.T) {
 	}()
 
-	expHash, err := testutil.DirHash(original)
+	expHash, err := dirHash(original)
```
Why not use `testutil.DirHash`?

krasi-georgiev replied:
ping @qiffang
